File: src/lib/wow/archive/casc/casc-source-remote.ts
/**
 * Remote (CDN) CASC source, ported from wow.export (src/js/casc/casc-source-remote.js).
 */
import util from 'util';

import { timeEnd, timeLog, write } from '@/lib/wow/log';
import { createProgress } from '@/lib/wow/server/runtime';

import { BufferWrapper } from '../../formats/buffer';
import { constants } from '../../formats/constants';
import {
  downloadFile, filesize, get, queue,
} from '../../formats/generics';
import { BLTEReader } from './blte-reader';
import { BuildCache } from './build-cache';
import type { CascKey } from './casc-key';
import { cascKeyFromHex, cascKeyToHex } from './casc-key';
import { CASC } from './casc-source';
import { CDNConfigEntries, parseCDNConfig } from './cdn-config';
import { cdnResolver } from './cdn-resolver';
import * as listfile from './listfile';
import { parseVersionConfig, VersionConfigEntry } from './version-config';

const EMPTY_HASH = '00000000000000000000000000000000';

interface ArchiveEntry {
  key: string;
  size: number;
  offset: number;
}

export class CASCRemote extends CASC {
  archives = new Map<string, ArchiveEntry>();

  region: string;

  host!: string;

  cache!: BuildCache;

  serverConfig!: VersionConfigEntry;

  /**
   * Create a new CASC source using a Blizzard CDN.
   * @param region Region tag (eu, us, etc).
   */
  constructor(region: string) {
    super(true);

    this.region = region;
  }

  /** Initialize remote CASC source. */
  async init(): Promise<void> {
    write('Initializing remote CASC source (%s)', this.region);
    this.host = util.format(constants.PATCH.HOST, this.region);
    this.builds = [];

    // Collect version configs for all products.
    const promises = constants.PRODUCTS.map((p) => this.getVersionConfig(p.product));
    const results = await Promise.allSettled(promises);

    // Iterate through successful requests and extract product config for our region.
    for (const result of results) {
      if (result.status === 'fulfilled') {
        const entry = result.value.find((e) => e.Region === this.region);
        if (entry) this.builds.push(entry);
      }
    }

    write('%o', this.builds);
  }

  /** Download the remote version config for a specific product. */
  async getVersionConfig(product: string): Promise<VersionConfigEntry[]> {
    const config = await this.getConfig(product, constants.PATCH.VERSION_CONFIG);
    config.forEach((entry) => { entry.Product = product; });
    return config;
  }

  /** Download and parse a version config file. */
  async getConfig(product: string, file: string): Promise<VersionConfigEntry[]> {
    const url = this.host + product + file;
    const res = await get(url);

    if (!res.ok) throw new Error(util.format('HTTP %d from remote CASC endpoint: %s', res.status, url));

    return parseVersionConfig(await res.text());
  }

  /**
   * Download and parse a CDN config file.
   * Attempts multiple CDN hosts in order of ping speed if one fails.
   * @param key
   * @param cdnHosts Optional array of CDN hosts to try (in priority order)
   */
  async getCDNConfig(key: string, cdnHosts: string[] | null = null): Promise<CDNConfigEntries> {
    // If no hosts provided, use the current host
    const hostsToTry = cdnHosts || [this.host];

    let lastError: Error | null = null;
    for (const host of hostsToTry) {
      try {
        const url = `${host}config/${this.formatCDNKey(key)}`;
        write('Attempting to retrieve CDN config from: %s', url);
        const res = await get(url);

        if (!res.ok) throw new Error(util.format('HTTP %d from CDN config endpoint', res.status));

        const configText = await res.text();
        const config = parseCDNConfig(configText);

        if (host !== this.host) {
          write('Successfully retrieved config from fallback host: %s', host);
          this.host = host;
        }

        return config;
      } catch (error) {
        write('Failed to retrieve CDN config from %s: %s', host, (error as Error).message);
        lastError = error as Error;

        cdnResolver.markHostFailed(host);
        continue;
      }
    }

    throw new Error(util.format('Unable to retrieve CDN config file %s from any CDN host. Last error: %s', key, lastError?.message || 'Unknown error'));
  }

  /** Obtain a file by its fileDataID. */
  async getFile(
    fileDataID: number,
    partialDecrypt = false,
    suppressLog = false,
    _supportFallback = true,
    _forceFallback = false,
    contentKey: CascKey | null = null,
  ): Promise<BLTEReader> {
    if (!suppressLog) write('Loading remote CASC file %d (%s)', fileDataID, listfile.getByID(fileDataID));

    const encodingKey = contentKey !== null ? this.getEncodingKeyForContentKey(contentKey) : this.getEncodingKey(fileDataID);
    const encodingKeyHex = cascKeyToHex(encodingKey);
    let data = await this.cache.getFile(encodingKeyHex, constants.CACHE.DIR_DATA);

    if (data === null) {
      const archive = this.archives.get(encodingKeyHex);
      if (archive !== undefined) {
        data = await this.getDataFilePartial(this.formatCDNKey(archive.key), archive.offset, archive.size);

        if (!suppressLog) write('Downloading CASC file %d from archive %s', fileDataID, archive.key);
      } else {
        data = await this.getDataFile(this.formatCDNKey(encodingKeyHex));

        if (!suppressLog) write('Downloading unarchived CASC file %d', fileDataID);

        if (data === null) throw new Error(`No remote unarchived/archive indexed for encoding key: ${encodingKeyHex}`);
      }

      void this.cache.storeFile(encodingKeyHex, data, constants.CACHE.DIR_DATA);
    } else if (!suppressLog) {
      write('Loaded CASC file %d from cache', fileDataID);
    }

    return new BLTEReader(data, encodingKeyHex, partialDecrypt);
  }

  /**
   * Returns a list of available products on the remote CDN.
   * Format example: "PTR: World of Warcraft 8.3.0.32272"
   */
  getProductList(): string[] {
    const products: string[] = [];
    for (const entry of this.builds) {
      // This check exists because some regions (e.g. China) may not have all products.
      if (entry === undefined) continue;

      const product = constants.PRODUCTS.find((e) => e.product === entry.Product)!;
      products.push(util.format('%s %s', product.title, entry.VersionsName));
    }

    return products;
  }

  /**
   * Preload requirements for reading remote files without initializing the
   * entire instance. Used by local CASC install for CDN fallback.
   */
  async preload(buildIndex: number, cache: BuildCache | null = null): Promise<void> {
    this.build = this.builds[buildIndex];
    write('Preloading remote CASC build: %o', this.build);

    if (cache) {
      this.cache = cache;
    } else {
      this.cache = new BuildCache(this.build.BuildConfig);
      await this.cache.init();
    }

    await this.loadServerConfig();
    await this.resolveCDNHost();
    await this.loadConfigs();
    await this.loadArchives();
  }

  /** Load the CASC interface with the given build. */
  async load(buildIndex: number): Promise<void> {
    this.resetForLoad();
    this.progress = createProgress(12);
    await this.preload(buildIndex);

    await this.loadEncoding();
    await this.loadRoot();

    await this.prepareListfile();
    await this.loadListfile(this.build.BuildConfig);

    await this.initializeComponents();
    this.isLoaded = true;
  }

  /** Download and parse the encoding file. */
  async loadEncoding(): Promise<void> {
    const encKeys = this.buildConfig.encoding.split(' ');
    const encKey = encKeys[1];

    timeLog();

    await this.progress.step('Loading encoding table');
    let encRaw = await this.cache.getFile(constants.CACHE.BUILD_ENCODING);
    if (encRaw === null) {
      // Encoding file not cached, download it.
      write('Encoding for build %s not cached, downloading.', this.cache.key);
      encRaw = await this.getDataFile(this.formatCDNKey(encKey));

      // Store back into cache (no need to block).
      void this.cache.storeFile(constants.CACHE.BUILD_ENCODING, encRaw);
    } else {
      write('Encoding for build %s cached locally.', this.cache.key);
    }

    timeEnd('Loaded encoding table (%s)', filesize(encRaw.byteLength));

    // Parse encoding file.
    timeLog();
    await this.progress.step('Parsing encoding table');
    this.parseEncodingFile(encRaw, encKey);
    timeEnd('Parsed encoding table (%d entries)', this.encodingKeys.size);
  }

  /** Download and parse the root file. */
  async loadRoot(): Promise<void> {
    // Get root key from encoding table.
    const rootKey = this.encodingKeys.get(cascKeyFromHex(this.buildConfig.root));
    if (rootKey === undefined) throw new Error('No encoding entry found for root key');

    timeLog();
    await this.progress.step('Loading root table');

    const rootKeyHex = cascKeyToHex(rootKey);
    let root = await this.cache.getFile(constants.CACHE.BUILD_ROOT);
    if (root === null) {
      // Root file not cached, download.
      write('Root file for build %s not cached, downloading.', this.cache.key);

      root = await this.getDataFile(this.formatCDNKey(rootKeyHex));
      void this.cache.storeFile(constants.CACHE.BUILD_ROOT, root);
    }

    timeEnd('Loaded root file (%s)', filesize(root.byteLength));

    // Parse root file.
    timeLog();
    await this.progress.step('Parsing root file');
    const rootEntryCount = this.parseRootFile(root, rootKeyHex);
    timeEnd('Parsed root file (%d entries, %d types)', rootEntryCount, this.rootTypes.length);
  }

  /** Download and parse archive files. */
  async loadArchives(): Promise<void> {
    // Download archive indexes.
    const archiveKeys = this.cdnConfig.archives.split(' ');
    const archiveCount = archiveKeys.length;

    timeLog();

    if (this.progress) await this.progress.step('Loading archives');

    await queue(archiveKeys, async (key: string) => this.parseArchiveIndex(key), 50);

    // Quick and dirty way to get the total archive size using config.
    const archiveTotalSize = this.cdnConfig.archivesIndexSize.split(' ').reduce((x, e) => Number(x) + Number(e), 0);
    timeEnd('Loaded %d archives (%d entries, %s)', archiveCount, this.archives.size, filesize(archiveTotalSize));
  }

  /**
   * Download the CDN configuration and store the entry for our selected region.
   */
  async loadServerConfig(): Promise<void> {
    if (this.progress) await this.progress.step('Fetching CDN configuration');

    // Download CDN server list.
    const serverConfigs = await this.getConfig(this.build.Product, constants.PATCH.SERVER_CONFIG);
    write('%o', serverConfigs);

    // Locate the CDN entry for our selected region.
    const serverConfig = serverConfigs.find((e) => e.Name === this.region);
    if (!serverConfig) throw new Error(`CDN config does not contain entry for region ${this.region}`);
    this.serverConfig = serverConfig;
  }

  /**
   * Load and parse the contents of an archive index.
   * Will use global cache and download if missing.
   */
  async parseArchiveIndex(key: string): Promise<void> {
    const fileName = `${key}.index`;

    let data = await this.cache.getFile(fileName, constants.CACHE.DIR_INDEXES);
    if (data === null) {
      const cdnKey = `${this.formatCDNKey(key)}.index`;
      data = await this.getDataFile(cdnKey);
      void this.cache.storeFile(fileName, data, constants.CACHE.DIR_INDEXES);
    }

    // Skip to the end of the archive to find the count.
    data.seek(-12);
    const count = data.readInt32LE();

    if (count * 24 > data.byteLength) throw new Error(`Unable to parse archive, unexpected size: ${data.byteLength}`);

    data.seek(0); // Reset position.

    for (let i = 0; i < count; i++) {
      let hash = data.readHexString(16);

      // Skip zero hashes.
      if (hash === EMPTY_HASH) hash = data.readHexString(16);

      this.archives.set(hash, { key, size: data.readInt32BE(), offset: data.readInt32BE() });
    }
  }

  /** Download a data file from the CDN. */
  async getDataFile(file: string): Promise<BufferWrapper> {
    return downloadFile(`${this.host}data/${file}`);
  }

  /** Download a partial chunk of a data file from the CDN. */
  async getDataFilePartial(file: string, ofs: number, len: number): Promise<BufferWrapper> {
    return downloadFile(`${this.host}data/${file}`, undefined, ofs, len);
  }

  /** Download the CDNConfig and BuildConfig. */
  async loadConfigs(): Promise<void> {
    if (this.progress) await this.progress.step('Fetching build configurations');

    const cdnHosts = await cdnResolver.getRankedHosts(this.region, this.serverConfig);

    this.cdnConfig = await this.getCDNConfig(this.build.CDNConfig, cdnHosts);
    this.buildConfig = await this.getCDNConfig(this.build.BuildConfig, cdnHosts);
  }

  /** Resolve the fastest CDN host for this region and server configuration. */
  async resolveCDNHost(): Promise<void> {
    if (this.progress) await this.progress.step('Locating fastest CDN server');

    this.host = await cdnResolver.getBestHost(this.region, this.serverConfig);
  }

  /**
   * Format a CDN key for use in CDN requests.
   * 49299eae4e3a195953764bb4adb3c91f -> 49/29/49299eae4e3a195953764bb4adb3c91f
   */
  formatCDNKey(key: string): string {
    return `${key.substring(0, 2)}/${key.substring(2, 4)}/${key}`;
  }

  /** Get the current build ID. */
  getBuildName(): string {
    return this.build.VersionsName;
  }

  /** Returns the build configuration key. */
  getBuildKey(): string {
    return this.build.BuildConfig;
  }
}

export default CASCRemote;

File: golang/internal/wow/archive/casc/casc_source_remote.go
package casc

import (
	"fmt"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/pqhuy98/wow-converter/internal/buffer"
	"github.com/pqhuy98/wow-converter/internal/wow/constants"
	"github.com/pqhuy98/wow-converter/internal/wow/formats"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
	"github.com/pqhuy98/wow-converter/internal/wow/runtime"
)

type archiveEntry struct {
	Key    string
	Size   int
	Offset int
}

// CASCRemote is a remote CDN CASC source.
type CASCRemote struct {
	*BaseCASC
	Archives     map[string]archiveEntry
	archivesMu   sync.Mutex
	Region       string
	Host         string
	ServerConfig VersionConfigEntry
	cdnHosts     []string
	cdnRetryAt   map[string]time.Time
	cdnHostsMu   sync.Mutex
}

// NewCASCRemote creates a remote CASC source.
func NewCASCRemote(region string) *CASCRemote {
	return &CASCRemote{
		BaseCASC:   NewBaseCASC(true),
		Archives:   map[string]archiveEntry{},
		Region:     region,
		cdnRetryAt: map[string]time.Time{},
	}
}

// Init initializes the remote CASC source.
func (c *CASCRemote) Init() error {
	log.Write("Initializing remote CASC source (%s)", c.Region)
	c.Host = fmt.Sprintf(constants.Patch.Host, c.Region)
	c.BuildsList = nil
	for _, product := range constants.Products {
		configs, err := c.GetVersionConfig(product.Product)
		if err != nil {
			continue
		}
		for _, entry := range configs {
			if entry["Region"] == c.Region {
				c.BuildsList = append(c.BuildsList, entry)
			}
		}
	}
	log.Write("%v", c.BuildsList)
	return nil
}

// GetVersionConfig downloads the remote version config for a product.
func (c *CASCRemote) GetVersionConfig(product string) ([]VersionConfigEntry, error) {
	config, err := c.GetConfig(product, constants.Patch.VersionConfig)
	if err != nil {
		return nil, err
	}
	for i := range config {
		config[i]["Product"] = product
	}
	return config, nil
}

// GetConfig downloads and parses a version config file.
func (c *CASCRemote) GetConfig(product, file string) ([]VersionConfigEntry, error) {
	url := c.Host + product + file
	res, err := formats.Get(url)
	if err != nil {
		return nil, err
	}
	if !res.OK {
		return nil, fmt.Errorf("HTTP %d from remote CASC endpoint: %s", res.Status, url)
	}
	return ParseVersionConfig(string(res.Body)), nil
}

// GetCDNConfig downloads and parses a CDN config file.
func (c *CASCRemote) GetCDNConfig(key string, cdnHosts []string) (CDNConfigEntries, error) {
	hostsToTry := cdnHosts
	if len(hostsToTry) == 0 {
		hostsToTry = []string{c.Host}
	}
	var lastError error
	for _, host := range hostsToTry {
		url := host + "config/" + c.FormatCDNKey(key)
		log.Write("Attempting to retrieve CDN config from: %s", url)
		res, err := formats.Get(url)
		if err != nil {
			lastError = err
			DefaultCDNResolver.MarkHostFailed(host)
			continue
		}
		if !res.OK {
			lastError = fmt.Errorf("HTTP %d from CDN config endpoint", res.Status)
			DefaultCDNResolver.MarkHostFailed(host)
			continue
		}
		cfg, err := ParseCDNConfig(string(res.Body))
		if err != nil {
			lastError = err
			DefaultCDNResolver.MarkHostFailed(host)
			continue
		}
		if host != c.Host {
			log.Write("Successfully retrieved config from fallback host: %s", host)
			c.Host = host
		}
		return cfg, nil
	}
	msg := "unknown error"
	if lastError != nil {
		msg = lastError.Error()
	}
	return nil, fmt.Errorf("unable to retrieve CDN config file %s from any CDN host. Last error: %s", key, msg)
}

// GetFile obtains a file by fileDataID.
func (c *CASCRemote) GetFile(fileDataID int, partialDecrypt, suppressLog, _, _ bool, contentKey CascKey) (*BLTEReader, error) {
	if !suppressLog {
		name, _ := GetByID(fileDataID)
		log.Write("Loading remote CASC file %d (%s)", fileDataID, name)
	}
	var encodingKey CascKey
	var err error
	if contentKey != "" {
		encodingKey, err = c.GetEncodingKeyForContentKey(contentKey)
	} else {
		encodingKey, err = c.GetEncodingKey(fileDataID)
	}
	if err != nil {
		return nil, err
	}
	encodingKeyHex := CascKeyToHex(encodingKey)
	data, _ := c.CacheStore.GetFile(encodingKeyHex, constants.Cache.DirData)
	if data == nil {
		archive, ok := c.Archives[encodingKeyHex]
		if ok {
			data, err = c.GetDataFilePartial(c.FormatCDNKey(archive.Key), archive.Offset, archive.Size)
			if !suppressLog {
				log.Write("Downloading CASC file %d from archive %s", fileDataID, archive.Key)
			}
		} else {
			data, err = c.GetDataFile(c.FormatCDNKey(encodingKeyHex))
			if !suppressLog {
				log.Write("Downloading unarchived CASC file %d", fileDataID)
			}
			if data == nil {
				return nil, fmt.Errorf("no remote unarchived/archive indexed for encoding key: %s", encodingKeyHex)
			}
		}
		if err != nil {
			return nil, err
		}
		go c.CacheStore.StoreFile(encodingKeyHex, data, constants.Cache.DirData)
	} else if !suppressLog {
		log.Write("Loaded CASC file %d from cache", fileDataID)
	}
	return NewBLTEReader(data, encodingKeyHex, partialDecrypt)
}

// GetProductList returns available products on the remote CDN.
func (c *CASCRemote) GetProductList() []string {
	var products []string
	for _, entry := range c.BuildsList {
		if entry == nil {
			continue
		}
		for _, product := range constants.Products {
			if product.Product == entry["Product"] {
				products = append(products, fmt.Sprintf("%s %s", product.Title, entry["VersionsName"]))
				break
			}
		}
	}
	return products
}

// Preload preloads requirements for CDN fallback usage.
func (c *CASCRemote) Preload(buildIndex int, cache *BuildCache) error {
	c.CurrentBuild = c.BuildsList[buildIndex]
	log.Write("Preloading remote CASC build: %v", c.CurrentBuild)
	if cache != nil {
		c.CacheStore = cache
	} else {
		c.CacheStore = NewBuildCache(c.CurrentBuild["BuildConfig"])
		if err := c.CacheStore.Init(); err != nil {
			return err
		}
	}
	if err := c.LoadServerConfig(); err != nil {
		return err
	}
	if err := c.ResolveCDNHost(); err != nil {
		return err
	}
	if err := c.LoadConfigs(); err != nil {
		return err
	}
	return c.LoadArchives()
}

// Load loads the CASC interface with the given build.
func (c *CASCRemote) Load(buildIndex int) error {
	c.ResetForLoad()
	c.ProgressTracker = runtime.CreateProgress(12)
	if err := c.Preload(buildIndex, nil); err != nil {
		return err
	}
	if err := c.LoadEncoding(); err != nil {
		return err
	}
	if err := c.LoadRoot(); err != nil {
		return err
	}
	if err := c.PrepareListfile(); err != nil {
		return err
	}
	if err := c.LoadListfile(); err != nil {
		return err
	}
	if err := c.InitializeComponents(); err != nil {
		return err
	}
	c.Loaded = true
	return nil
}

// LoadEncoding downloads and parses the encoding file.
func (c *CASCRemote) LoadEncoding() error {
	encKeys := strings.Fields(c.BuildConfigData["encoding"])
	encKey := encKeys[1]
	log.TimeLog()
	if err := c.ProgressTracker.Step("Loading encoding table"); err != nil {
		return err
	}
	encRaw, _ := c.CacheStore.GetFile(constants.Cache.BuildEncoding, "")
	if encRaw == nil {
		log.Write("Encoding for build %s not cached, downloading.", c.CacheStore.Key)
		var err error
		encRaw, err = c.GetDataFile(c.FormatCDNKey(encKey))
		if err != nil {
			return err
		}
		go c.CacheStore.StoreFile(constants.Cache.BuildEncoding, encRaw, "")
	} else {
		log.Write("Encoding for build %s cached locally.", c.CacheStore.Key)
	}
	log.TimeEnd("Loaded encoding table (%s)", formats.Filesize(encRaw.ByteLength()))
	log.TimeLog()
	if err := c.ProgressTracker.Step("Parsing encoding table"); err != nil {
		return err
	}
	if err := c.ParseEncodingFile(encRaw, encKey); err != nil {
		return err
	}
	log.TimeEnd("Parsed encoding table (%d entries)", c.EncodingTable.len())
	return nil
}

// LoadRoot downloads and parses the root file.
func (c *CASCRemote) LoadRoot() error {
	rootKey, ok := c.EncodingTable.lookupEncodingKey(CascKeyFromHex(c.BuildConfigData["root"]))
	if !ok {
		return fmt.Errorf("no encoding entry found for root key")
	}
	log.TimeLog()
	if err := c.ProgressTracker.Step("Loading root table"); err != nil {
		return err
	}
	rootKeyHex := CascKeyToHex(rootKey)
	root, _ := c.CacheStore.GetFile(constants.Cache.BuildRoot, "")
	if root == nil {
		log.Write("Root file for build %s not cached, downloading.", c.CacheStore.Key)
		var err error
		root, err = c.GetDataFile(c.FormatCDNKey(rootKeyHex))
		if err != nil {
			return err
		}
		go c.CacheStore.StoreFile(constants.Cache.BuildRoot, root, "")
	}
	log.TimeEnd("Loaded root file (%s)", formats.Filesize(root.ByteLength()))
	log.TimeLog()
	if err := c.ProgressTracker.Step("Parsing root file"); err != nil {
		return err
	}
	rootEntryCount, err := c.ParseRootFile(root, rootKeyHex)
	if err != nil {
		return err
	}
	log.TimeEnd("Parsed root file (%d entries, %d types)", rootEntryCount, len(c.RootTypesList))
	return nil
}

// LoadArchives downloads and parses archive files.
func (c *CASCRemote) LoadArchives() error {
	archiveKeys := strings.Fields(c.CDNConfigData["archives"])
	archiveCount := len(archiveKeys)
	log.TimeLog()
	if c.ProgressTracker != nil {
		if err := c.ProgressTracker.Step("Loading archives"); err != nil {
			return err
		}
	}
	if err := formats.Queue(archiveKeys, func(key string) error {
		return c.ParseArchiveIndex(key)
	}, 50); err != nil {
		return err
	}
	archiveTotalSize := 0
	for _, e := range strings.Fields(c.CDNConfigData["archivesIndexSize"]) {
		n, _ := strconv.Atoi(e)
		archiveTotalSize += n
	}
	log.TimeEnd("Loaded %d archives (%d entries, %s)", archiveCount, len(c.Archives), formats.Filesize(archiveTotalSize))
	return nil
}

// LoadServerConfig downloads CDN configuration for the selected region.
func (c *CASCRemote) LoadServerConfig() error {
	if c.ProgressTracker != nil {
		if err := c.ProgressTracker.Step("Fetching CDN configuration"); err != nil {
			return err
		}
	}
	serverConfigs, err := c.GetConfig(c.CurrentBuild["Product"], constants.Patch.ServerConfig)
	if err != nil {
		return err
	}
	log.Write("%v", serverConfigs)
	for _, entry := range serverConfigs {
		if entry["Name"] == c.Region {
			c.ServerConfig = entry
			return nil
		}
	}
	return fmt.Errorf("CDN config does not contain entry for region %s", c.Region)
}

// ParseArchiveIndex loads and parses an archive index.
func (c *CASCRemote) ParseArchiveIndex(key string) error {
	fileName := key + ".index"
	data, _ := c.CacheStore.GetFile(fileName, constants.Cache.DirIndexes)
	if data == nil {
		cdnKey := c.FormatCDNKey(key) + ".index"
		var err error
		data, err = c.GetDataFile(cdnKey)
		if err != nil {
			return err
		}
		go c.CacheStore.StoreFile(fileName, data, constants.Cache.DirIndexes)
	}
	data.Seek(-12)
	count := int(data.ReadInt32LE().(int64))
	if count*24 > data.ByteLength() {
		return fmt.Errorf("unable to parse archive, unexpected size: %d", data.ByteLength())
	}
	data.Seek(0)
	entries := make(map[string]archiveEntry, count)
	for i := 0; i < count; i++ {
		hash := data.ReadHexString(16)
		if hash == emptyHash {
			hash = data.ReadHexString(16)
		}
		entries[hash] = archiveEntry{
			Key:    key,
			Size:   int(data.ReadInt32BE().(int64)),
			Offset: int(data.ReadInt32BE().(int64)),
		}
	}
	c.archivesMu.Lock()
	for hash, entry := range entries {
		c.Archives[hash] = entry
	}
	c.archivesMu.Unlock()
	return nil
}

// GetDataFile downloads a data file from the CDN.
func (c *CASCRemote) GetDataFile(file string) (*buffer.Buffer, error) {
	return c.downloadDataFile(file, -1, -1)
}

// GetDataFilePartial downloads a partial chunk of a data file.
func (c *CASCRemote) GetDataFilePartial(file string, ofs, length int) (*buffer.Buffer, error) {
	return c.downloadDataFile(file, ofs, length)
}

func (c *CASCRemote) downloadDataFile(file string, ofs, length int) (*buffer.Buffer, error) {
	var lastErr error
	for _, host := range c.healthyCDNHosts(time.Now()) {
		data, err := formats.DownloadFile([]string{host + "data/" + file}, "", ofs, length, false)
		if err == nil {
			return data, nil
		}
		lastErr = err
		c.deferCDNHost(host, time.Now())
		log.Write("CASC data host failed, trying fallback: %s (%s)", host, err)
	}
	return nil, lastErr
}

func (c *CASCRemote) healthyCDNHosts(now time.Time) []string {
	c.cdnHostsMu.Lock()
	defer c.cdnHostsMu.Unlock()

	hosts := c.cdnHosts
	if len(hosts) == 0 {
		hosts = []string{c.Host}
	}
	ready := make([]string, 0, len(hosts))
	deferred := make([]string, 0, len(hosts))
	for _, host := range hosts {
		if retryAt := c.cdnRetryAt[host]; retryAt.After(now) {
			deferred = append(deferred, host)
		} else {
			ready = append(ready, host)
		}
	}
	if len(ready) == 0 {
		return deferred
	}
	return append(ready, deferred...)
}

func (c *CASCRemote) deferCDNHost(host string, now time.Time) {
	c.cdnHostsMu.Lock()
	defer c.cdnHostsMu.Unlock()
	c.cdnRetryAt[host] = now.Add(30 * time.Second)
}

// LoadConfigs downloads CDNConfig and BuildConfig.
func (c *CASCRemote) LoadConfigs() error {
	if c.ProgressTracker != nil {
		if err := c.ProgressTracker.Step("Fetching build configurations"); err != nil {
			return err
		}
	}
	cdnHosts, err := DefaultCDNResolver.GetRankedHosts(c.Region, c.ServerConfig)
	if err != nil {
		return err
	}
	c.CDNConfigData, err = c.GetCDNConfig(c.CurrentBuild["CDNConfig"], cdnHosts)
	if err != nil {
		return err
	}
	c.BuildConfigData, err = c.GetCDNConfig(c.CurrentBuild["BuildConfig"], cdnHosts)
	return err
}

// ResolveCDNHost resolves the fastest CDN host.
func (c *CASCRemote) ResolveCDNHost() error {
	if c.ProgressTracker != nil {
		if err := c.ProgressTracker.Step("Locating fastest CDN server"); err != nil {
			return err
		}
	}
	hosts, err := DefaultCDNResolver.GetRankedHosts(c.Region, c.ServerConfig)
	if err != nil {
		return err
	}
	c.cdnHostsMu.Lock()
	c.cdnHosts = hosts
	c.cdnRetryAt = map[string]time.Time{}
	c.Host = hosts[0]
	c.cdnHostsMu.Unlock()
	return nil
}

// FormatCDNKey formats a CDN key for CDN requests.
func (c *CASCRemote) FormatCDNKey(key string) string {
	return key[:2] + "/" + key[2:4] + "/" + key
}

// GetBuildName returns the current build ID.
func (c *CASCRemote) GetBuildName() string { return c.CurrentBuild["VersionsName"] }

// GetBuildKey returns the build configuration key.
func (c *CASCRemote) GetBuildKey() string { return c.CurrentBuild["BuildConfig"] }
